File: 1. Fundamental Level/1.1 Advanced C#/1. Advanced-CSharp-Arrays-Lists-Stacks-Queues-Homework/LongestIncreasingSequence.cs
/*
 * Problem 5.	Longest Increasing Sequence
 * Write a program to find all increasing sequences inside an array of integers.
 * The integers are given on a single line, separated by a space.
 * Print the sequences in the order of their appearance in the input array, each at a single line. 
 * Separate the sequence elements by a space. 
 * Find also the longest increasing sequence and print it at the last line. 
 * If several sequences have the same longest length, print the left-most of them. 
 * Examples:
 * 
 * Input	            Output
 * 2 3 4 1 50 2 3 4 5	2 3 4
 *                      1 50
 *                      2 3 4 5
 *                      Longest: 2 3 4 5
 */
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

class LongestIncreasingSequence
{
    static void Main(string[] args)
    {
        Console.Write("Enter numbers: ");
        string readNumbers = Console.ReadLine();

        char[] separator = new char[] { ' ' };
        List<int> numbers = readNumbers.Split(separator, StringSplitOptions.RemoveEmptyEntries).Select(Int32.Parse).ToList();
        PrintIncrSequences(numbers);

    }

    public static void PrintIncrSequences(List<int> numbers)
    {
        int length = numbers.Count;
        int number, nextNumber = 0, count = 1, bestCount = -1;
        string result = null, bestResult = null;

        for (int i = 0; i < length; i++)
        {
            number = numbers[i];

            if (i != length - 1)
            {
                nextNumber = numbers[i + 1];
            }

            if(number < nextNumber)
            {
                Console.Write(number + " ");
                count++;
                result += number + " ";
            }
            else
            {
                Console.WriteLine(number);
                result += number + " ";
                count++;

                if (count > bestCount)
                {
                    bestCount = count;
                    bestResult = result;
                }

                count = 1;
                result = null;
            }
        }

        Console.WriteLine("Longest: " + bestResult);
    }
}

File: 1. Fundamental Level/1.4 Java Basics/6. Java-Collections-Basics-Homework/LongestIncreasingSequence.java
/*
 * Problem 4. Longest Increasing Sequence
 * Write a program to find all increasing sequences inside an array of integers. 
 * The integers are given in a single line, separated by a space. 
 * Print the sequences in the order of their appearance in the input array, each at a single line. 
 * Separate the sequence elements by a space. 
 * Find also the longest increasing sequence and print it at the last line. 
 * If several sequences have the same longest length, print the leftmost of them.
 * */
import java.util.ArrayList;
import java.util.Scanner;

public class LongestIncreasingSequence {
	public static ArrayList<Integer> numbers;
	public static int length;
	public static int number;
	
	public static void main(String[] args) {		
		ArrayList<Integer> numbers = readNumbers();
		findIncreasingSequence(numbers);
	}
	
	/**
	 * Read numbers.
	 *
	 * @return the array list
	 */
	public static ArrayList<Integer> readNumbers() {
		numbers = new ArrayList<Integer>();
		Scanner scanner = new Scanner(System.in);
		
		System.out.print("Enter numbers: ");
		String readNumbers = scanner.nextLine();
		String[] splitNumbers = readNumbers.split(" ");
		length = splitNumbers.length;
		
		for (int i = 0; i < length; i++) {
			number = Integer.parseInt(splitNumbers[i]);
			numbers.add(number);
		}
		
		scanner.close();
		
		return numbers;
	}
	
	/**
	 * Find increasing sequence.
	 *
	 * @param numbers the numbers
	 */
	public static void findIncreasingSequence(ArrayList<Integer> numbers) {
		length = numbers.size();
		int nextNumber = 0, count = 1, bestCount = -1;
		String result = "", bestResult = null;
		
		for (int i = 0; i < length; i++) {
			number = numbers.get(i);
			
			if(i != length - 1) {
				nextNumber = numbers.get(i + 1);
			}
			
			if(number < nextNumber) {
				System.out.printf("%d ", number);
				count++;
				result += number + " ";
			} else {
				System.out.println(number);
				result += number + " ";
				count++;
				
				if(count > bestCount) {
					bestCount = count;
					bestResult = result;
				}
				
				count = 1;
				result = "";
			}
		}
		
		
		System.out.printf("Longest: %s\n", bestResult);
	}
}
